sdk/go/pulumi/types.go

```diff
@@ -240,6 +262,13 @@ func NewOutput() (Output, func(interface{}), func(error)) {
 	return AnyOutput{out}, resolve, reject
 }
 
+// NewOutput returns an output value that can be used to rendezvous with the production of a value or error.  The
+// function returns the output itself, plus two functions: one for resolving a value, and another for rejecting with an
+// error; exactly one function must be called. This acts like a promise.
+func NewOutput() (Output, func(interface{}), func(error)) {
```

Comment:
Do we want to augment the doc comments for these to mention the new methods on Context? Are the Context methods now preferable, and if so, are we eventually going to mark these as deprecated? Should we just go ahead and do that now?

Reply:
> if so, are we eventually going to mark these as deprecated? Should we just go ahead and do that now?

I think we should, yes (meant to leave a review comment to that effect, d'oh). @lukehoban wdyt?

Reply:
Which methods exactly are moving?

Is it all the global methods here including `All`?  Some of these are very heavily used in typical Pulumi code, and I'm not sure it's "better" to ask users to write the alternative using context?  What does some real code look like after this change?  What exactly goes wrong if users continue to use the existing functions?

Reply:
I think moving all the global methods would make sense, or at least the combinators. From a developer perspective, having them be context-less introduces the potential to wonder about if it's possible to pass in multiple outputs with different contexts and what underlying context the combinator then uses. When, in reality, I can't think of any scenario where one would do something like that and would expect all outputs to share the same context as the combinator. Moving the combinators to the context-level would make it fairly unambiguous that the combinator operates purely within that context. One could make a similar argument for the other methods (e.g. does `ToSecret` have some context that's different from the underlying `Output`'s?) so I would argue for moving them all to hang off context as opposed to leaving them at the top-level.

Reply:
One interesting data point I just encountered: we use `pulumi.All` inside the implementation of the `Index`/`MapIndex` methods that we generate for array and map outputs. These methods do not have direct access to a `pulumi.Context`, though they do have indirect access via the receiver (since the receiver is an output). This pushes me a bit towards the side of only moving `NewOutput` onto `pulumi.Context`. From a technical standpoint, we could leave the combinators (`All`, `Any`, `ToOutput`, `ToSecret` and others we might eventually add) at the top-level and have them continue to take the strategy implemented in this PR of digging the appropriate wait group out of any output arguments.

Reply:
That would scope the deprecations to only `NewOutput`.

Note that in general users should not be using combinators on outputs associated with different contexts. The only way to do that would be to escape outputs from multiple contexts via shared state. These changes do handle that situation as correctly as possible by bumping the outstanding work counter on all contexts, but it is very much an edge case.